### ISSUE TITLE
Different round/cast orders in MatExpr, depending on seemingly unrelated compiler flags

### ISSUE THREAD

**Issue by kamjagin** [opencv/opencv#7144]
##### System information (version)

- OpenCV => 3.0 (tested on the 3.0.0 tag and master) 
- Operating System / Platform => OSX10.11 and Ubuntu16.04
- Compiler => clang-703.0.29 and gcc-5.4.0

##### Detailed description

It seems that there is some random (undefined?) behaviour when a type conversion occurs in a `MatExpr`. Sometimes the cast (`convertTo`) happens before the multiplication operation (and therefore no precision loss because of rounding) and sometimes the cast happens after.

So `cv::Mat_<float> = cv::Mat_<uint16_t> * float` (pseudocode) produces different results depending on seemingly unrelated compiler flags in the target app (i.e. OpenCV compilation is unchanged for the two different results).

##### Steps to reproduce
Compile opencv fresh from master.

**reproduce_conversion_fail.cpp**:
```
#include <iostream>
#include <string>

#include <opencv2/opencv.hpp>

int main()
{
    cv::Mat_<uint16_t> imu = cv::Mat_<uint16_t>::ones(10, 10);

    cv::Mat_<uint16_t> imu2 = imu * 0.001f;
    imu2 *= 1000.f;

    cv::Mat_<float> imf = imu * 0.001f;
    imf *= 1000.f;

    std::cout << "mean uint16_t: " << cv::mean(imu) << std::end;
    std::cout << "mean uint16_t (autorounded): " << cv::mean(imu2) << std::end;
    std::cout << "mean float: " << cv::mean(imf) << std::endl;
}
```

`g++ reproduce_conversion_fail.cpp -I../src/opencv/install/include -L../src/opencv/install/lib -lopencv_core -lopencv_imgcodecs -std=c++11` results in:
```
mean uint16_t: [1, 0, 0, 0]
mean uint16_t (autorounded): [0, 0, 0, 0]
mean float: [0, 0, 0, 0]
```
or
`g++ reproduce_conversion_fail.cpp -I../src/opencv/install/include -L../src/opencv/install/lib -lopencv_core -lopencv_imgcodecs` results in
```
mean uint16_t: [1, 0, 0, 0]
mean uint16_t (autorounded): [0, 0, 0, 0]
mean float: [1, 0, 0, 0]
```

Compiling opencv with `WITH_IPP=OFF` removes this behaviour. Not sure if the error is in the IPP-part of the code or if it just cannot be triggered with the `c++11` flag when IPP is enabled.

**Comment by jamestautges** [comment]
Couldn't replicate the issue with gcc-6.1, but I'd like to keep investigating this issue.

**Comment by alalek** [comment]
Actually this statement deals wrong. It should not mix types:
```
cv::Mat_<float> imf = imu * 0.001f;
```

Element-wise multiplication is performed and stored with "uint16_t" datatype. And only after this result is converted to float.

You should use explicit data conversion like this:
```
    cv::Mat_<float> imu_float = imu;
    cv::Mat_<float> imf = imu_float * 0.001f;
```

**Comment by kamjagin** [comment]
@alalek I disagree with the removal of the bug label: If implicit conversion is known to produce random behaviour, we should make sure that an assert of some kind is raised when a user of the library tries to do it. If on the other hand implicit conversion is supported (given that you are fully aware of the truncation that will happen since the multiplication is performed with the uint16_t datatype), the results should be predictable and not dependent on seemingly unrelated compiler flags (as indicated in the original post).

**Comment by bedbad** [comment]
Can't reproduce with any flags on newer versions of gcc.
It is no issue with clang either. I don't see how it is opencv related.
